File: src/framing.rs
// SPDX-License-Identifier: Apache-2.0

use prost::Message;

use crate::heddle::api::v1alpha1::{CallContext, CallFailure};

/// Largest fully-qualified method path accepted by the hosted-call protocol.
pub const MAX_METHOD_PATH: usize = 1024;
/// Largest encoded call context accepted before dispatch.
pub const MAX_CALL_CONTEXT: usize = 64 * 1024;
/// Largest protobuf control body carried in a FIN-delimited frame.
pub const MAX_CONTROL_BODY: usize = 8 * 1024 * 1024;

const RESPONSE_SUCCESS: u8 = 0;
const RESPONSE_FAILURE: u8 = 1;
const STREAM_MESSAGE: u8 = 0;
const STREAM_FAILURE: u8 = 1;
const STREAM_HEADER: usize = 5;

/// Malformed or oversized hosted-call framing.
#[derive(Debug, thiserror::Error)]
pub enum FrameError {
    /// A length or path violates the contract ceiling.
    #[error("invalid hosted-call frame: {0}")]
    Invalid(String),
    /// A protobuf context or failure envelope could not be decoded.
    #[error("invalid hosted-call protobuf: {0}")]
    Decode(#[from] prost::DecodeError),
}

/// Decoded request whose body remains borrowed from the FIN-delimited frame.
#[derive(Debug)]
pub struct RequestFrame<'a> {
    /// Canonical fully-qualified method path.
    pub method: &'a str,
    /// Typed call metadata decoded before routing.
    pub context: CallContext,
    /// Encoded method request body.
    pub body: &'a [u8],
}

/// Decoded unary response outcome.
#[derive(Debug)]
pub enum ResponseFrame<'a> {
    /// Encoded successful response body.
    Success(&'a [u8]),
    /// Contract-owned failure envelope.
    Failure(CallFailure),
}

/// One length-delimited item within a server or bidirectional stream.
#[derive(Debug)]
pub enum StreamFrame<'a> {
    /// Encoded protobuf stream message.
    Message(&'a [u8]),
    /// Terminal contract-owned failure.
    Failure(CallFailure),
}

/// Encodes `method_len:u16be | context_len:u32be | method | context | body`.
/// The operation stream FIN is the outer delimiter.
pub fn encode_request_frame(
    method: &str,
    context: &CallContext,
    body: &[u8],
) -> Result<Vec<u8>, FrameError> {
    validate_method(method)?;
    validate_body(body)?;
    let context = context.encode_to_vec();
    if context.len() > MAX_CALL_CONTEXT {
        return Err(FrameError::Invalid(format!(
            "call context is {} bytes; maximum is {MAX_CALL_CONTEXT}",
            context.len()
        )));
    }
    let method_len = u16::try_from(method.len())
        .map_err(|_| FrameError::Invalid("method path exceeds u16".to_string()))?;
    let context_len = u32::try_from(context.len())
        .map_err(|_| FrameError::Invalid("call context exceeds u32".to_string()))?;
    let mut frame = Vec::with_capacity(6 + method.len() + context.len() + body.len());
    frame.extend_from_slice(&method_len.to_be_bytes());
    frame.extend_from_slice(&context_len.to_be_bytes());
    frame.extend_from_slice(method.as_bytes());
    frame.extend_from_slice(&context);
    frame.extend_from_slice(body);
    Ok(frame)
}

/// Decodes a complete FIN-delimited request frame.
pub fn decode_request_frame(frame: &[u8]) -> Result<RequestFrame<'_>, FrameError> {
    if frame.len() < 6 {
        return Err(FrameError::Invalid(
            "request prelude is truncated".to_string(),
        ));
    }
    let method_len = u16::from_be_bytes([frame[0], frame[1]]) as usize;
    let context_len = u32::from_be_bytes([frame[2], frame[3], frame[4], frame[5]]) as usize;
    if method_len == 0 || method_len > MAX_METHOD_PATH || context_len > MAX_CALL_CONTEXT {
        return Err(FrameError::Invalid(
            "request prelude declares an invalid length".to_string(),
        ));
    }
    let context_start = 6_usize
        .checked_add(method_len)
        .ok_or_else(|| FrameError::Invalid("request length overflow".to_string()))?;
    let body_start = context_start
        .checked_add(context_len)
        .ok_or_else(|| FrameError::Invalid("request length overflow".to_string()))?;
    if body_start > frame.len() {
        return Err(FrameError::Invalid(
            "request frame is truncated".to_string(),
        ));
    }
    let method = std::str::from_utf8(&frame[6..context_start])
        .map_err(|_| FrameError::Invalid("method path is not UTF-8".to_string()))?;
    validate_method(method)?;
    let body = &frame[body_start..];
    validate_body(body)?;
    Ok(RequestFrame {
        method,
        context: CallContext::decode(&frame[context_start..body_start])?,
        body,
    })
}

/// Encodes a successful unary response; stream FIN delimits the body.
pub fn encode_success_response(body: &[u8]) -> Result<Vec<u8>, FrameError> {
    validate_body(body)?;
    let mut frame = Vec::with_capacity(1 + body.len());
    frame.push(RESPONSE_SUCCESS);
    frame.extend_from_slice(body);
    Ok(frame)
}

/// Encodes a contract-owned unary failure; stream FIN delimits the envelope.
pub fn encode_failure_response(failure: &CallFailure) -> Result<Vec<u8>, FrameError> {
    let body = failure.encode_to_vec();
    validate_body(&body)?;
    let mut frame = Vec::with_capacity(1 + body.len());
    frame.push(RESPONSE_FAILURE);
    frame.extend_from_slice(&body);
    Ok(frame)
}

/// Decodes a complete FIN-delimited unary response frame.
pub fn decode_response_frame(frame: &[u8]) -> Result<ResponseFrame<'_>, FrameError> {
    let (&outcome, body) = frame
        .split_first()
        .ok_or_else(|| FrameError::Invalid("response frame is empty".to_string()))?;
    validate_body(body)?;
    match outcome {
        RESPONSE_SUCCESS => Ok(ResponseFrame::Success(body)),
        RESPONSE_FAILURE => Ok(ResponseFrame::Failure(CallFailure::decode(body)?)),
        value => Err(FrameError::Invalid(format!(
            "unknown response outcome {value}"
        ))),
    }
}

/// Encodes one protobuf message for a streaming operation.
pub fn encode_stream_message(body: &[u8]) -> Result<Vec<u8>, FrameError> {
    encode_stream_item(STREAM_MESSAGE, body)
}

/// Encodes one terminal failure for a streaming operation.
pub fn encode_stream_failure(failure: &CallFailure) -> Result<Vec<u8>, FrameError> {
    encode_stream_item(STREAM_FAILURE, &failure.encode_to_vec())
}

/// Decodes one item from a streaming receive buffer.
///
/// Returns `Ok(None)` until the buffer contains the complete declared item.
/// The consumed length lets callers retain any following frames already read.
pub fn decode_stream_frame(buffer: &[u8]) -> Result<Option<(StreamFrame<'_>, usize)>, FrameError> {
    if buffer.len() < STREAM_HEADER {
        return Ok(None);
    }
    let kind = buffer[0];
    let body_len = u32::from_be_bytes([buffer[1], buffer[2], buffer[3], buffer[4]]) as usize;
    if body_len > MAX_CONTROL_BODY {
        return Err(FrameError::Invalid(format!(
            "stream item is {body_len} bytes; maximum is {MAX_CONTROL_BODY}"
        )));
    }
    let consumed = STREAM_HEADER
        .checked_add(body_len)
        .ok_or_else(|| FrameError::Invalid("stream item length overflow".to_string()))?;
    if buffer.len() < consumed {
        return Ok(None);
    }
    let body = &buffer[STREAM_HEADER..consumed];
    let frame = match kind {
        STREAM_MESSAGE => StreamFrame::Message(body),
        STREAM_FAILURE => StreamFrame::Failure(CallFailure::decode(body)?),
        value => {
            return Err(FrameError::Invalid(format!(
                "unknown stream item kind {value}"
            )));
        }
    };
    Ok(Some((frame, consumed)))
}

fn encode_stream_item(kind: u8, body: &[u8]) -> Result<Vec<u8>, FrameError> {
    validate_body(body)?;
    let body_len = u32::try_from(body.len())
        .map_err(|_| FrameError::Invalid("stream item exceeds u32".to_string()))?;
    let mut frame = Vec::with_capacity(STREAM_HEADER + body.len());
    frame.push(kind);
    frame.extend_from_slice(&body_len.to_be_bytes());
    frame.extend_from_slice(body);
    Ok(frame)
}

fn validate_method(method: &str) -> Result<(), FrameError> {
    if method.is_empty() || !method.starts_with('/') || method.len() > MAX_METHOD_PATH {
        return Err(FrameError::Invalid(
            "method path must begin with '/' and fit the method-path limit".to_string(),
        ));
    }
    Ok(())
}

fn validate_body(body: &[u8]) -> Result<(), FrameError> {
    if body.len() > MAX_CONTROL_BODY {
        return Err(FrameError::Invalid(format!(
            "control body is {} bytes; maximum is {MAX_CONTROL_BODY}",
            body.len()
        )));
    }
    Ok(())
}

File: packages/typescript/runtime/framing.ts
// SPDX-License-Identifier: Apache-2.0
//
// Hosted-call frame codec, byte-compatible with the authoritative Rust
// implementation in `src/framing.rs` (crate `heddle-api`). This is the wire
// format the hosted-call ALPN carries over one bidirectional operation stream:
// the stream FIN delimits each control frame.
//
// Request frame:  method_len:u16be | context_len:u32be | method (utf8)
//                 | CallContext (protobuf) | body
// Response frame: outcome:u8 (0 = success, 1 = failure) | body
//                 (a failure body is a CallFailure protobuf envelope)
//
// Only the unary request/response frames are modelled here; that is the shape
// the ADR-0049 browser claim transport uses (see heddle's claim_protocol.rs,
// which frames with encode_request_frame + decode_response_frame).

import { fromBinary, toBinary } from "@bufbuild/protobuf";
import { CallContextSchema, type CallContext } from "./contract_pb.js";
import { CallFailureSchema, type CallFailure } from "./errors_pb.js";

/** Largest fully-qualified method path accepted by the hosted-call protocol. */
export const MAX_METHOD_PATH = 1024;
/** Largest encoded call context accepted before dispatch. */
export const MAX_CALL_CONTEXT = 64 * 1024;
/** Largest protobuf control body carried in a FIN-delimited frame. */
export const MAX_CONTROL_BODY = 8 * 1024 * 1024;

const RESPONSE_SUCCESS = 0;
const RESPONSE_FAILURE = 1;
const PRELUDE_HEADER = 6;

/** Malformed or oversized hosted-call framing. */
export class FrameError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "FrameError";
  }
}

/** Decoded request frame; `body` is a view over the supplied frame bytes. */
export interface RequestFrame {
  /** Canonical fully-qualified method path. */
  method: string;
  /** Typed call metadata decoded before routing. */
  context: CallContext;
  /** Encoded method request body. */
  body: Uint8Array;
}

/** Decoded unary response outcome. */
export type ResponseFrame =
  | { readonly outcome: "success"; readonly body: Uint8Array }
  | { readonly outcome: "failure"; readonly failure: CallFailure };

const textEncoder = new TextEncoder();
// `fatal` rejects non-UTF-8 method paths, matching Rust's `str::from_utf8`.
const textDecoder = new TextDecoder("utf-8", { fatal: true });

function validateBodyLen(bodyLen: number): void {
  if (bodyLen > MAX_CONTROL_BODY) {
    throw new FrameError(
      `control body is ${bodyLen} bytes; maximum is ${MAX_CONTROL_BODY}`,
    );
  }
}

function validateMethodBytes(method: string, methodByteLen: number): void {
  if (method.length === 0 || !method.startsWith("/") || methodByteLen > MAX_METHOD_PATH) {
    throw new FrameError(
      "method path must begin with '/' and fit the method-path limit",
    );
  }
}

/**
 * Encodes `method_len:u16be | context_len:u32be | method | context | body`.
 * The operation stream FIN is the outer delimiter.
 */
export function encodeRequestFrame(
  method: string,
  context: CallContext,
  body: Uint8Array,
): Uint8Array {
  validateBodyLen(body.length);
  const prelude = encodeRequestPrelude(method, context);
  const frame = new Uint8Array(prelude.length + body.length);
  frame.set(prelude, 0);
  frame.set(body, prelude.length);
  return frame;
}

/** Encodes the method and typed context that precede a request body. */
export function encodeRequestPrelude(
  method: string,
  context: CallContext,
): Uint8Array {
  const methodBytes = textEncoder.encode(method);
  validateMethodBytes(method, methodBytes.length);
  const contextBytes = toBinary(CallContextSchema, context);
  if (contextBytes.length > MAX_CALL_CONTEXT) {
    throw new FrameError(
      `call context is ${contextBytes.length} bytes; maximum is ${MAX_CALL_CONTEXT}`,
    );
  }
  // `methodBytes.length <= MAX_METHOD_PATH` guarantees the u16, and the context
  // ceiling guarantees the u32; the explicit guards mirror the Rust try_from.
  if (methodBytes.length > 0xffff) {
    throw new FrameError("method path exceeds u16");
  }
  if (contextBytes.length > 0xffff_ffff) {
    throw new FrameError("call context exceeds u32");
  }

  const frame = new Uint8Array(PRELUDE_HEADER + methodBytes.length + contextBytes.length);
  const view = new DataView(frame.buffer);
  view.setUint16(0, methodBytes.length, false);
  view.setUint32(2, contextBytes.length, false);
  frame.set(methodBytes, PRELUDE_HEADER);
  frame.set(contextBytes, PRELUDE_HEADER + methodBytes.length);
  return frame;
}

/** Decodes a complete FIN-delimited request frame. */
export function decodeRequestFrame(frame: Uint8Array): RequestFrame {
  if (frame.length < PRELUDE_HEADER) {
    throw new FrameError("request frame contains a truncated prelude");
  }
  const view = new DataView(frame.buffer, frame.byteOffset, frame.byteLength);
  const methodLen = view.getUint16(0, false);
  const contextLen = view.getUint32(2, false);
  if (methodLen === 0 || methodLen > MAX_METHOD_PATH || contextLen > MAX_CALL_CONTEXT) {
    throw new FrameError("request prelude declares an invalid length");
  }
  const contextStart = PRELUDE_HEADER + methodLen;
  const consumed = contextStart + contextLen;
  if (frame.length < consumed) {
    throw new FrameError("request frame contains a truncated prelude");
  }

  let method: string;
  try {
    method = textDecoder.decode(frame.subarray(PRELUDE_HEADER, contextStart));
  } catch {
    throw new FrameError("method path is not UTF-8");
  }
  validateMethodBytes(method, methodLen);

  const context = decodeCallContext(frame.subarray(contextStart, consumed));
  const body = frame.subarray(consumed);
  validateBodyLen(body.length);
  return { method, context, body };
}

/** Encodes a successful unary response; the stream FIN delimits the body. */
export function encodeSuccessResponse(body: Uint8Array): Uint8Array {
  validateBodyLen(body.length);
  const frame = new Uint8Array(1 + body.length);
  frame[0] = RESPONSE_SUCCESS;
  frame.set(body, 1);
  return frame;
}

/** Encodes a contract-owned unary failure; the stream FIN delimits the envelope. */
export function encodeFailureResponse(failure: CallFailure): Uint8Array {
  const body = toBinary(CallFailureSchema, failure);
  validateBodyLen(body.length);
  const frame = new Uint8Array(1 + body.length);
  frame[0] = RESPONSE_FAILURE;
  frame.set(body, 1);
  return frame;
}

/** Decodes a complete FIN-delimited unary response frame. */
export function decodeResponseFrame(frame: Uint8Array): ResponseFrame {
  if (frame.length === 0) {
    throw new FrameError("response frame is empty");
  }
  const outcome = frame[0];
  const body = frame.subarray(1);
  validateBodyLen(body.length);
  switch (outcome) {
    case RESPONSE_SUCCESS:
      return { outcome: "success", body };
    case RESPONSE_FAILURE:
      return { outcome: "failure", failure: decodeCallFailure(body) };
    default:
      throw new FrameError(`unknown response outcome ${outcome}`);
  }
}

function decodeCallContext(bytes: Uint8Array): CallContext {
  try {
    return fromBinary(CallContextSchema, bytes);
  } catch (cause) {
    throw new FrameError(`invalid hosted-call protobuf: ${String(cause)}`);
  }
}

function decodeCallFailure(bytes: Uint8Array): CallFailure {
  try {
    return fromBinary(CallFailureSchema, bytes);
  } catch (cause) {
    throw new FrameError(`invalid hosted-call protobuf: ${String(cause)}`);
  }
}
